src/inputs.rs

```diff
@@ -107,228 +135,291 @@ impl<'source> FromPyObject<'source> for PartitionedInput {
-            // behavior, but also means we don't have to find a way to
-            // propogate the correct partition:worker mappings into the
-            // restore system, which would be more difficult as we have to
-            // find a way to treat this kind of state key differently. I
-            // might regret this.
-            .filter(|key| key.is_local(index, worker_count))
-            .flat_map(|key| {
-                let state = resume_state.remove(&key).map(StateBytes::de::<TdPyAny>);
-                tracing::info!(
-                    "{index:?} building input {step_id:?} \
-                    source {key:?} with resume state {state:?}"
-                );
-                match self
-                    .0
-                    .call_method1(py, "build_part", (key.clone(), state))
-                    .and_then(|part| part.extract(py))
-                {
-                    Err(err) => Some(Err(err)),
-                    Ok(None) => None,
-                    Ok(Some(part)) => Some(Ok((key, part))),
-                }
-            })
-            .collect::<PyResult<HashMap<StateKey, StatefulSource>>>()
-            .reraise("error creating input source partitions")?;
-
-        if !resume_state.is_empty() {
-            tracing::warn!(
-                "Resume state exists for {step_id:?} \
-                for unknown partitions {:?}; \
-                changing partition counts? \
-                recovery state routing bug?",
-                resume_state.keys()
-            );
-        }
-
-        Ok(parts)
     }
 
     /// Read items from a partitioned input.
     ///
-    /// This is a stateful operator, so the change stream must be
-    /// incorporated into the recovery system, and the resume state
-    /// must be routed back here.
+    /// Will manage assigning primary workers for all partitions and
+    /// building them.
     ///
-    /// Will manage automatically distributing partition sources. All
-    /// you have to do is pass in the definition.
+    /// This can't be unified into the recovery system input operators
+    /// because they are stateless and have no epoch semantics.
     pub(crate) fn partitioned_input<S>(
         self,
         py: Python,
-        scope: &S,
+        scope: &mut S,
         step_id: StepId,
         epoch_interval: EpochInterval,
-        index: WorkerIndex,
-        count: WorkerCount,
         probe: &ProbeHandle<u64>,
         start_at: ResumeEpoch,
-        resume_state: StepStateBytes,
-    ) -> PyResult<(Stream<S, TdPyAny>, FlowChangeStream<S>)>
+        loads: &Stream<S, Snapshot>,
+    ) -> PyResult<(Stream<S, TdPyAny>, Stream<S, Snapshot>)>
     where
         S: Scope<Timestamp = u64>,
     {
-        let mut parts = self.build(py, step_id.clone(), index, count, resume_state)?;
+        let this_worker = scope.w_index();
 
-        let mut op_builder = OperatorBuilder::new(step_id.0.clone(), scope.clone());
+        let local_parts = self.list_parts(py)?;
```

Comment:
I would add a `.reraise("...")` here, so that we get the line info where this failed. What I currently see is:
```
thread '<unnamed>' panicked at 'Box<dyn Any>', src/run.rs:117:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/docler/src/bytewax/bytewax/pysrc/bytewax/run.py", line 422, in <module>
    cli_main(**kwargs)
TypeError: (src/worker.rs:152:10) error building production dataflow
Caused by => TypeError: (src/worker.rs:292:30) error building PartitionedInput
Caused by => TypeError: 'set' object cannot be converted to 'Sequence'
```
And as you can see, the error only points us up to `worker.rs`, and we can't see where the error happened, while adding a `reraise` here gives us:
```
thread '<unnamed>' panicked at 'Box<dyn Any>', src/run.rs:117:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/docler/src/bytewax/bytewax/pysrc/bytewax/run.py", line 422, in <module>
    cli_main(**kwargs)
TypeError: (src/worker.rs:152:10) error building production dataflow
Caused by => TypeError: (src/worker.rs:292:30) error building PartitionedInput
Caused by => TypeError: (src/inputs.rs:187:47) error listing parts
Caused by => TypeError: 'set' object cannot be converted to 'Sequence'
```